enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,284 @@
+
+### Non-Goals
+- Stateful support
+- Flood Attacks
+
+## Proposal
+
+- customers will label nodes where ingress node policy will be deployed on.
+- customers will deploy daemonSet controller on all the selected nodes by applying `IngressNodeFirewallConfig` object. 
+- customers will prepare `IngressNodeFirewall` objects with all required firewall rules and interfaces the ebpf program will be attached to.
+- IngressNodeFW controller will process `IngressNodeFirewall` objects, verification webhook will make sure configuration doesn't contradict with fail safe rules.
+- Ingress node firewall daemon will watch ingress node firewall objects and create `nodeState` object which contains all rules objects that is applied to specific node then load eBPF map table and attach eBPF kernel hook loaded as `BPF_PROG_TYPE_XDP` to the list of interfaces configured in the object.
+- in order to fail safe and avoid cluster lockup, various pinholes will be open in the firewall to allow basic services.
+
+### Workflow Description
+
+- the following diagram shows how node ingress firewall configuration is handled
+
+![Ingress Firewall configuration processing](./ingress-nodefw-config-flow.png)
+
+- the following diagram shows eBPF how ingress firewall kernel hook will process packets and apply firewall rules
+
+![eBPF workflow](./ebpf-packet-flow.png)
+
+- the following table shows sample of fail safe rules to make sure users won't block basic cluster services by configuring invalid firewall rules
+
+| Source CIDR | Protocol | Port | Purpose |
+|-------------|----------|------|---------|
+|  0.0.0.0/0  |   TCP    |  22  |  SSH    |
+|  0.0.0.0/0  |   UDP    |  68  | DHCP    |
+|  0.0.0.0/0  |   TCP    | 6443 | Kubernetes API server access |
+|  0.0.0.0/0  |   TCP    | 2380 & 2379 | Etcd access |
+
+### Implementation Details/Notes/Constraints
+
+A new downstream OLM operator (openshift/ingress-nodefw-operator) will be designed to do the following:
+- Create Ingress node firewall config object to deploy daemonSet.
+- Create the CRDs for the new object.
+- controller for each of the new resources to handle add/update/delete
+- verification webhook to validate the configuration and enforce failSafe rules.
+- library to construct and load eBPF ingress maps tables based on the provisioned rules.
+- library for statistics collection and exporting to prometheus.
+- library for bpf perf event reporting.
+- library for FailSafe rules handling.
+- library to sync eBPF rules and update rules in eBPF maps.
+
+New downstream repo for the kernel hook implementation (openshift/ingress-node-firewall)
+- sample `IngressNodeConfig` object
+  
+```yaml
+apiVersion: ingressnodefirewall.openshift.io/v1alpha1
+kind: IngressNodeFirewallConfig
+metadata:
+  name: ingressnodefirewallconfig
+  namespace: ingress-node-firewall-system
+spec:
+  nodeSelector:
+    node-role.kubernetes.io/worker: ""
+  tolerations:
+  - key: "Example"
+    operator: "Exists"
+    effect: "NoExecute"
```

Comment:
(As above, I don't think we should have the `IngressNodeFirewallConfig`, but if we do then you shouldn't have to manually specify `tolerations` in it.)

Reply:
can u explain what is the prob setting toleration for it ?

Reply:
Setting tolerations on the DaemonSet is fine, I'm just saying that we should do that _automatically_ rather than requiring the user to do it explicitly.

Reply:
In that case they don't want the software to be able to run on every node, so they allow the admin to clarify which kinds of nodes the software can run on and which ones it can't. In our case, if there are firewall rules that apply to the node, and there is any software running on that node at all, then the daemon _must_ run on that node.

Reply:
should i dropped from the config object then ?